python/ray/util/client/common.py

```diff
@@ -83,9 +84,167 @@
+
+    def __del__(self):
+        if self._worker is not None and self._worker.is_connected():
+            try:
+                if not self.is_nil():
+                    self._worker.call_release(self.id)
+            except Exception:
+                logger.info(
+                    "Exception in ObjectRef is ignored in destructor. "
+                    "To receive this exception in application code, call "
+                    "a method on the actor reference before its destructor "
+                    "is run."
+                )
+
+    def binary(self):
+        self._wait_for_id()
+        return super().binary()
+
+    def hex(self):
+        self._wait_for_id()
+        return super().hex()
+
+    def is_nil(self):
+        self._wait_for_id()
+        return super().is_nil()
+
+    def __hash__(self):
+        self._wait_for_id()
+        return hash(self.id)
+
+    def task_id(self):
+        self._wait_for_id()
+        return super().task_id()
+
+    @property
+    def id(self):
+        return self.binary()
+
+    def future(self) -> Future:
+        fut = Future()
+
+        def set_future(data: Any) -> None:
+            """Schedules a callback to set the exception or result
+            in the Future."""
+
+            if isinstance(data, Exception):
+                fut.set_exception(data)
+            else:
+                fut.set_result(data)
+
+        self._on_completed(set_future)
+
+        # Prevent this object ref from being released.
+        fut.object_ref = self
+        return fut
+
+    def _on_completed(self, py_callback: Callable[[Any], None]) -> None:
+        """Register a callback that will be called after Object is ready.
+        If the ObjectRef is already ready, the callback will be called soon.
+        The callback should take the result as the only argument. The result
+        can be an exception object in case of task error.
+        """
+
+        def deserialize_obj(
+            resp: Union[ray_client_pb2.DataResponse, Exception]
+        ) -> None:
+            from ray.util.client.client_pickler import loads_from_server
+
+            if isinstance(resp, Exception):
+                data = resp
+            else:
+                obj = resp.get
+                data = None
+                if not obj.valid:
+                    data = loads_from_server(resp.get.error)
+                else:
+                    data = loads_from_server(resp.get.data)
+
+            py_callback(data)
+
+        self._worker.register_callback(self, deserialize_obj)
+
+    def _set_id(self, id):
+        super()._set_id(id)
+        self._worker.call_retain(id)
+
+    def _wait_for_id(self, timeout=None):
+        if self._id_future:
+            with self._mutex:
+                if self._id_future:
+                    self._set_id(self._id_future.result(timeout=timeout))
+                    self._id_future = None
+
+
+class ClientActorRef(raylet.ActorID):
```

Comment:
Interesting, I thought we had these in Cython originally so that we could extend the proper Cython types as sub-classes. So this is unnecessary?

Reply:
I think we first have them in python, and because of `_set_id` need to access `self.data`, it's moved to cython.
I'm not 100% about the story here. Maybe @mwtian can share some thoughts.